### PR TITLE
Implementation of statistic functions for an admin

### LINKED ISSUE #47: Implement Statistic Manager
This manager will permit to ask the server for statistics such as stats for a Discussion(time, number of participants etc..) or Questions(number of responses, interest etc..)

### PR DESCRIPTION
**Pull request by Naludrag**
The admin can now ask for stats of the debates.
fixes #47

### CODE REVIEW

File: src/namespace/privilegednamespace.js [hunk 1]
```diff
@@ -213,4 +219,54 @@ export class PrivilegedNamespace extends CustomNamespace {
...
+    /**
+     * Return an array that contains stats for the debates of an admin in the result of the callback function
+     */
+    getAdminStats = (socket) => async (callback) => {
+        logger.debug(`getAdminStats received from ${socket.id}`);
+
+        if (!(callback instanceof Function)) {
+            logger.debug(`callback is not a function.`);
+            return;
+        }
```
Comment (ghost) [inline src/namespace/privilegednamespace.js:232]: I see this block of code appearing quite often. Is it possible to refactor it? Like extracting a method. 

You don't have to apply this change now, I will still approve this PR anyways if there is no other problem with it.

File: src/namespace/privilegednamespace.js [hunk 2]
```diff
@@ -213,4 +219,54 @@ export class PrivilegedNamespace extends CustomNamespace {
...
+    /**
+     * Return an array that contains stats for a specific debate in the result of the callback function
+     */
+    getDebateStats = (socket) => async (debateId, callback) => {
+        logger.debug(`getDebatStats received from ${socket.id}`);
+
+        if (!(callback instanceof Function)) {
+            logger.debug(`callback is not a function.`);
+            return;
+        }
```
Comment (ghost) [inline src/namespace/privilegednamespace.js:255]: You see?

File: src/namespace/privilegednamespace.js [hunk 3]
```diff
@@ -213,4 +219,54 @@ export class PrivilegedNamespace extends CustomNamespace {
...
+    /**
+     * Return an array that contains stats for a specific question in the result of the callback function
+     */
+    getQuestionStats = (socket) => async (questionId, debateId, callback) => {
+        logger.debug(`getQuestionStats received from ${socket.id}`);
+
+        if (!(callback instanceof Function)) {
+            logger.debug(`callback is not a function.`);
+            return;
+        }
```
Comment (ghost) [inline src/namespace/privilegednamespace.js:277]: You see?

File: src/statistic/statistic.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Get the number of the total of votes for all the responses for a question.
+     * @param aQuestionId integer that is the id of the question
+     * @param aDiscussionId integer that is the id of the discussion
+     * @returns {Promise<number>} an integer that is the number of votes
+     */
+    async getNumberVotesQuestion(aQuestionId, aDiscussionId) {
+        // Get all the responses related to the question desired
+        let allResponses = await dbManager.getResponsesQuestion(aQuestionId, aDiscussionId);
```
Comment (ghost) [inline src/statistic/statistic.js:87]: Can `dbManager.getResponsesQuestion` return a null object ? If that's the case we need to verify the value of `allResponses` before using it.

File: src/statistic/statistic.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Get all the stats for a debate
+     * @param aDiscussionId the id of the discussion to get the stats from
+     * @returns {Promise<*[]>} an array containing the stats wanted for a debate
+     */
+    async debateStats(aDiscussionId){
+        // Get the discussion in the database
+        let debate = await dbManager.getDiscussion(aDiscussionId);
+        // Get all the questions related to the debate
+        let allQuestions = await dbManager.getQuestionsDiscussion(aDiscussionId);
```
Comment (ghost) [inline src/statistic/statistic.js]: Same here. Can `dbManager.getDiscussion` or `dbManager.getQuestionsDiscussion` return a null object ?

File: src/statistic/statistic.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Get all the stats a the question
+     * @param aQuestionId integer that is the id of the question
+     * @param aDiscussionId integer that is the id of the discussion
+     * @returns {Promise<*[]>} Array that contains the stats wanted for a question
+     */
+   async questionStats(aQuestionId, aDiscussionId){
+       let debate = await dbManager.getDiscussion(aDiscussionId);
+       let allResponses = await dbManager.getResponsesQuestion(aQuestionId, aDiscussionId);
+       let numberTotalVotes = await this.getNumberVotesQuestion(aQuestionId,aDiscussionId);
```
Comment (ghost) [inline src/statistic/statistic.js]: Same here. Can `dbManager.getDiscussion`, `dbManager.getResponsesQuestion` or `this.getNumberVotesQuestion` return a null object ?

File: test/30_debate/adminstatisticfunctions.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    describe("Get stats", () => {
+        before(async ()  => {
+            debateManager = new DebateManager();
+            await debateManager.start();
+
+            admin = io.connect(`${PRIVILEGED_NAMESPACE}`, {
+                path: SocketConfig.DEFAULT_PATH,
+                forceNew: true,
+                query: {
+                    password: `${SocketConfig.ADMIN_PASSWORD}`,
```
Comment (ghost) [inline test/30_debate/adminstatisticfunctions.js:25]: CWE 798 ( _L O L_ )

File: src/namespace/privilegednamespace.js (hunk 1 above)
Comment (ghost) [reply src/namespace/privilegednamespace.js:232]: The block of code I'm talking about (it wasn't appearing correctly on my view)
```js
if (!(callback instanceof Function)) {
    logger.debug(`callback is not a function.`);
    return;
}
```

Comment (Naludrag) [reply src/namespace/privilegednamespace.js:232]: I will try to change it in the next PR

File: src/statistic/statistic.js (hunk 4 above)
Comment (Naludrag) [reply src/statistic/statistic.js:87]: Yes it was effectively a case that I didn't implement I changed it in the latest commit

File: src/statistic/statistic.js (hunk 5 above)
Comment (Naludrag) [reply src/statistic/statistic.js]: As said before corrected in the new commit

File: src/statistic/statistic.js (hunk 6 above)
Comment (Naludrag) [reply src/statistic/statistic.js]: As said before corrected in the new commit

File: test/30_debate/adminstatisticfunctions.js (hunk 7 above)
Comment (Naludrag) [reply test/30_debate/adminstatisticfunctions.js:25]: Corrected it is now a password of 256 non-ASCII characters and I'm the only one that knows it

**Review by ghost — APPROVED** [review]

**Pull request merged**